src/import-utils.ts

```diff
@@ -0,0 +1,301 @@
+import {ASTNode} from 'parse5';
+import {Analyzer, Options as AnalyzerOptions} from 'polymer-analyzer';
+import {Document, ScannedDocument} from 'polymer-analyzer/lib/model/document';
+import {Import} from 'polymer-analyzer/lib/model/import';
+import {ParsedHtmlDocument} from 'polymer-analyzer/lib/html/html-document';
+import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
+import constants from './constants';
+import * as astUtils from './ast-utils';
+import * as matchers from './matchers';
+import * as urlUtils from './url-utils';
+import {Bundle, BundleStrategy, AssignedBundle, generateBundles, BundleUrlMapper, BundleManifest, sharedBundleUrlMapper, generateSharedDepsMergeStrategy} from './bundle-manifest';
+import {BundledDocument, DocumentCollection} from './document-collection';
+import {buildDepsIndex} from './deps-index';
+import {UrlString} from './url-utils';
+
+// TODO(usergenic): Want to figure out a way to get rid of the basePath param
+// used in the inline functions, because it feels obnoxious to have to pass it
+// around for an only-occasionally used case.
+
+/**
+ * Inline the contents of the html document returned by the link tag's href
+ * at the location of the link tag and then remove the link tag.
+ */
+export async function inlineHtmlImport(
+    basePath: string|undefined,
+    docUrl: string,
+    htmlImport: ASTNode,
+    reachedImports: Set<string>,
+    docBundle: AssignedBundle,
+    manifest: BundleManifest,
+    loader: (url: UrlString) => Promise<string>) {
+  const rawImportUrl: string = dom5.getAttribute(htmlImport, 'href')!;
+  const resolvedImportUrl: string = urlLib.resolve(docUrl, rawImportUrl);
+  const importBundleUrl = manifest.bundleUrlForFile.get(resolvedImportUrl);
+
+  // Don't reprocess the same file again.
+  if (reachedImports.has(resolvedImportUrl)) {
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // If we can't find a bundle for the referenced import, record that we've
+  // processed it, but don't remove the import link.  Browser will handle it.
+  if (!importBundleUrl) {
+    reachedImports.add(resolvedImportUrl);
+    return;
+  }
+
+  // Don't inline an import into itself.
+  if (docUrl === resolvedImportUrl) {
+    reachedImports.add(resolvedImportUrl);
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // Guard against inlining a import we've already processed.
+  if (reachedImports.has(importBundleUrl)) {
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // If the html import refers to a file which is bundled and has a different
+  // url, then lets just rewrite the href to point to the bundle url.
+  if (importBundleUrl !== docBundle.url) {
+    const relative =
+        urlUtils.relativeUrl(docUrl, importBundleUrl) || importBundleUrl;
+    dom5.setAttribute(htmlImport, 'href', relative);
+    reachedImports.add(importBundleUrl);
+    return;
+  }
+
+  const document =
```

Comment:
Added `no-unused-variable: true` in 01403af55fc28482904e91fb5180bbdaa4c33767 and it caught a lot of things.  Not as many as I expected, but a few unused functions and such were pruned.  Lots of unused imports too of course.